src/source/src/row_id.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::{SystemTime, UNIX_EPOCH};
+
+const TIMESTAMP_SHIFT_BITS: u8 = 22;
+const WORKER_ID_SHIFT_BITS: u8 = 12;
+const SEQUENCE_UPPER_BOUND: u16 = 1 << 12;
+const WORKER_ID_UPPER_BOUND: u32 = 1 << 10;
+
+/// `RowIdGenerator` generates unique row ids using snowflake algorithm as following format:
+///
+/// | timestamp | worker id | sequence |
+/// |-----------|-----------|----------|
+/// |  41 bits  | 10 bits   | 12 bits  |
+#[derive(Debug)]
+pub struct RowIdGenerator {
+    /// Specific epoch using for generating row ids.
+    epoch: SystemTime,
```

Comment:
Will this time go backward? For example, after some automatic NTP syncing.

Reply:
There is only one possible case where time fallback needs to be considered, and that is when the compute node is restarted. I have discussed this with @fuyufjh , just "Bai" it now.